`text/plain` and `openmetrics` parsers panic on unquoted metric name inside braces (e.g. `{A}0`), crashing the collector

### What did you do?

The Prometheus text format and OpenMetrics parsers panic when they encounter an
entry whose line starts with `{` and contains a bare (unquoted) identifier not
followed by `=`, such as `{A}0`. Because the scrape pipeline has no panic
recovery, this crashes the entire Prometheus process.

**Affected parsers:** `PromParser` (`text/plain`) and `OpenMetricsParser`
(`application/openmetrics-text`)

## Steps to reproduce

Stand up any HTTP endpoint returning this body with `Content-Type: text/plain`:

```
{A}0
```

Point a Prometheus instance at it. The collector crashes on the first scrape.

The panic can also be reproduced in isolation:

```
package main

import (
	"github.com/prometheus/prometheus/model/labels"
	"github.com/prometheus/prometheus/model/textparse"
)

// Run against unpatched code: panic: runtime error: slice bounds out of range [2:1]
func main() {
	p, _ := textparse.New([]byte("{A} 0\n"), "application/openmetrics-text", nil, textparse.ParserOptions{})
	p.Next()
	var lset labels.Labels
	p.Labels(&lset)
}
```

```
package main

import (
	"github.com/prometheus/prometheus/model/labels"
	"github.com/prometheus/prometheus/model/textparse"
)

// Run against unpatched code: panic: runtime error: slice bounds out of range [2:1]
func main() {
	p, _ := textparse.New([]byte("{A}0\n"), "text/plain", nil, textparse.ParserOptions{})
	p.Next()
	var lset labels.Labels
	p.Labels(&lset)
}
```


### What did you expect to see?

The parser returns a parse error from `Next()` and the scrape loop logs it and
continues normally, as it does for any other malformed metric line.


### What did you see instead? Under which circumstances?

`Next()` returns `(EntrySeries, nil)` — no error — because the corrupt offsets
are not detected at parse time. The panic is deferred to the subsequent
`Labels()` call:

```
goroutine N [running]:
github.com/prometheus/prometheus/model/textparse.(*PromParser).Labels(...)
    model/textparse/promparse.go:237
panic: runtime error: slice bounds out of range [2:1]
```


### Prometheus version

I'm able to reproduce this on `main`.

--- a/model/textparse/openmetricsparse.go
+++ b/model/textparse/openmetricsparse.go
@@ -634,11 +634,13 @@ func (p *OpenMetricsParser) parseLVals(offsets []int, isExemplar bool) ([]int, e
 	for {
 		curTStart := p.l.start
 		curTI := p.l.i
+		var isQString bool
 		switch t {
 		case tBraceClose:
 			return offsets, nil
 		case tLName:
 		case tQString:
+			isQString = true
 		default:
 			return nil, p.parseError("expected label name", t)
 		}
@@ -647,7 +649,7 @@ func (p *OpenMetricsParser) parseLVals(offsets []int, isExemplar bool) ([]int, e
 		// A quoted string followed by a comma or brace is a metric name. Set the
 		// offsets and continue processing. If this is an exemplar, this format
 		// is not allowed.
-		if t == tComma || t == tBraceClose {
+		if isQString && (t == tComma || t == tBraceClose) {
 			if isExemplar {
 				return nil, p.parseError("expected label name", t)
 			}

--- a/model/textparse/openmetricsparse_test.go
+++ b/model/textparse/openmetricsparse_test.go
@@ -942,6 +942,10 @@ func TestOpenMetricsParseErrors(t *testing.T) {
 			err:   "expected label name, got \"=\\\"\" (\"EQUAL\") while parsing: \"empty_label_name{=\\\"\"",
 		},
 		{
+			input: "{A}0\n# EOF\n",
+			err:   "expected equal, got \"}0\" (\"BCLOSE\") while parsing: \"{A}0\"",
+		},
+		{
 			input: "foo 1_2\n\n# EOF\n",
 			err:   "unsupported character in float while parsing: \"foo 1_2\"",
 		},
@@ -971,11 +975,11 @@ func TestOpenMetricsParseErrors(t *testing.T) {
 		},
 		{
 			input: `custom_metric_total 1 # {bb}`,
-			err:   "expected label name, got \"}\" (\"BCLOSE\") while parsing: \"custom_metric_total 1 # {bb}\"",
+			err:   "expected equal, got \"}\" (\"BCLOSE\") while parsing: \"custom_metric_total 1 # {bb}\"",
 		},
 		{
 			input: `custom_metric_total 1 # {bb, a="dd"}`,
-			err:   "expected label name, got \", \" (\"COMMA\") while parsing: \"custom_metric_total 1 # {bb, \"",
+			err:   "expected equal, got \", \" (\"COMMA\") while parsing: \"custom_metric_total 1 # {bb, \"",
 		},
 		{
 			input: `custom_metric_total 1 # {aa="bb",,cc="dd"} 1`,
@@ -1035,6 +1039,22 @@ func TestOpenMetricsParseErrors(t *testing.T) {
 		}
 		require.Equal(t, c.err, err.Error(), "test %d: %s", i, c.input)
 	}
+}
+
+func TestOpenMetricsParseBareIdentifierInBraces(t *testing.T) {
+	require.NotPanics(t, func() {
+		p := NewOpenMetricsParser([]byte("{A} 0\n# EOF\n"), labels.NewSymbolTable(), WithOMParserSTSeriesSkipped())
+		for {
+			et, err := p.Next()
+			if err != nil {
+				break
+			}
+			if et == EntrySeries {
+				var lset labels.Labels
+				p.Labels(&lset)
+			}
+		}
+	})
 }
 
 func TestOMNullByteHandling(t *testing.T) {

--- a/model/textparse/promparse.go
+++ b/model/textparse/promparse.go
@@ -408,11 +408,13 @@ func (p *PromParser) parseLVals() error {
 	for {
 		curTStart := p.l.start
 		curTI := p.l.i
+		var isQString bool
 		switch t {
 		case tBraceClose:
 			return nil
 		case tLName:
 		case tQString:
+			isQString = true
 		default:
 			return p.parseError("expected label name", t)
 		}
@@ -420,7 +422,7 @@ func (p *PromParser) parseLVals() error {
 		t = p.nextToken()
 		// A quoted string followed by a comma or brace is a metric name. Set the
 		// offsets and continue processing.
-		if t == tComma || t == tBraceClose {
+		if isQString && (t == tComma || t == tBraceClose) {
 			if p.offsets[0] != -1 || p.offsets[1] != -1 {
 				return fmt.Errorf("metric name already set while parsing: %q", p.l.b[p.start:p.l.i])
 			}

--- a/model/textparse/promparse_test.go
+++ b/model/textparse/promparse_test.go
@@ -511,6 +511,10 @@ func TestPromParseErrors(t *testing.T) {
 			err:   "expected label name, got \"=\\\"\" (\"EQUAL\") while parsing: \"empty_label_name{=\\\"\"",
 		},
 		{
+			input: "{A}0",
+			err:   "expected equal, got \"}0\" (\"BCLOSE\") while parsing: \"{A}0\"",
+		},
+		{
 			input: "foo 1_2\n",
 			err:   "unsupported character in float while parsing: \"foo 1_2\"",
 		},
@@ -548,6 +552,22 @@ func TestPromParseErrors(t *testing.T) {
 		}
 		require.EqualError(t, err, c.err, "test %d", i)
 	}
+}
+
+func TestPromParseBareIdentifierInBraces(t *testing.T) {
+	require.NotPanics(t, func() {
+		p := NewPromParser([]byte("{A} 0\n"), labels.NewSymbolTable(), false)
+		for {
+			et, err := p.Next()
+			if err != nil {
+				break
+			}
+			if et == EntrySeries {
+				var lset labels.Labels
+				p.Labels(&lset)
+			}
+		}
+	})
 }
 
 func TestPromNullByteHandling(t *testing.T) {

